### PR TITLE
Add submission download flag for staff users

### PR DESCRIPTION
**Pull request by ghost**
Add checkmark next to submission filename to indicate that the submission has been downloaded (via the filename link or batch download).

### CODE REVIEW

File: edx_sga/sga.py [hunk 1]
```diff
@@ -256,16 +257,16 @@ def get_student_data():
...
                     'student_id': student.student_id,
                     'submission_id': submission['uuid'],
                     'username': module.student.username,
                     'fullname': module.student.profile.name,
                     'filename': submission['answer']["filename"],
-                    'timestamp': submission['created_at'].strftime(
-                        DateTime.DATETIME_FORMAT
-                    ),
+                    'downloaded': downloaded,
+                    'timestamp': submission['created_at'].strftime('%m/%d/%Y %-I:%M%p').lower(),
```
Comment (iivic) [inline edx_sga/sga.py]: Why did you change the format?

File: edx_sga/static/js/src/edx_sga.js [hunk 2]
```diff
@@ -142,6 +142,15 @@ function StaffGradedAssignmentXBlock(runtime, element) {
                 });
             });
 
+            // Set up submission download links
+            var submissionDownloadedHtml = "<i class=\"fa fa-check-circle\" " +
```
Comment (iivic) [inline edx_sga/static/js/src/edx_sga.js]: I would move variable definitions to top of the function.

File: edx_sga/static/js/src/edx_sga.js [hunk 3]
```diff
@@ -142,6 +142,15 @@ function StaffGradedAssignmentXBlock(runtime, element) {
...
+                "aria-hidden=\"true\" title=\"Submission downloaded\"></i>";
+            $("a.submission-filename").on("click", function () {
```
Comment (iivic) [inline edx_sga/static/js/src/edx_sga.js]: Selectors shouldn't contain HTML tags.

Comment (iivic) [reply edx_sga/static/js/src/edx_sga.js]: There can be multiple xblock instances in the same page. To prevent using the wrong element use `$(".submission-filename", element)`.

File: edx_sga/static/js/src/edx_sga.js [hunk 4]
```diff
@@ -142,6 +142,15 @@ function StaffGradedAssignmentXBlock(runtime, element) {
...
+                if (!$(this).prev("i").length) {
+                   $(this).before(submissionDownloadedHtml);
+                }
```
Comment (iivic) [inline edx_sga/static/js/src/edx_sga.js]: This is hard to understand. Can you rewrite it so your intentions are easy to understand or add a comment which will explain it?

File: edx_sga/static/js/src/edx_sga.js [hunk 5]
```diff
@@ -161,13 +170,22 @@ function StaffGradedAssignmentXBlock(runtime, element) {
             $("button#download-selected-submissions").on("click", function() {
                 var student_ids = [];
                 $submissionCheckboxes.filter(":checked").each(function() {
+                    var $submissionFilename = $(this).closest("tr").find("a.submission-filename");
```
Comment (iivic) [inline edx_sga/static/js/src/edx_sga.js]: Once again without HTML tag names.

File: edx_sga/static/js/src/edx_sga.js [hunk 6]
```diff
@@ -161,13 +170,22 @@ function StaffGradedAssignmentXBlock(runtime, element) {
...
+                        $submissionFilename.before(submissionDownloadedHtml);
+                    }
                     student_ids.push($(this).val());
                 });
                 $.post(downloadSubmissionsUrl, JSON.stringify({'student_ids': student_ids})).done(function(data) {
                     window.location = window.location.origin + data.zip_url;
                 });
             });
             $("button#download-all-submissions").on("click", function() {
+                $("a.submission-filename").each(function() {
```
Comment (iivic) [inline edx_sga/static/js/src/edx_sga.js]: Without `a` tag and use `$(".submission-filename", element)`.

**Review by iivic — CHANGES_REQUESTED** [review]

File: edx_sga/sga.py [hunk 7]
```diff
@@ -491,6 +495,16 @@ def download_zip(self, student_files):
         # save zip file and return its URL as JSON response
         return Response(json={'zip_url': default_storage.url(default_storage.save(zip_filename, sio))})
 
+    def get_submission_download_status(self, student_id):
+        student_item_dict = self.student_submission_id(student_id)
+        staff_user = User.objects.get(id=self.xmodule_runtime.user_id)
+        return submissions_api.get_download_status(student_item_dict, staff_user)
+
+    def set_submission_as_downloaded(self, student_id):
```
Comment (marjev) [inline edx_sga/sga.py]: Nitpick: I think renaming this to `set_submission_status_to_downloaded` makes more sense

File: edx_sga/sga.py (hunk 1 above)
Comment (marjev) [inline edx_sga/sga.py]: Save the timestamp as a constant, preferably within the XBlock class.

File: edx_sga/sga.py [hunk 8]
```diff
@@ -491,6 +495,16 @@ def download_zip(self, student_files):
         # save zip file and return its URL as JSON response
         return Response(json={'zip_url': default_storage.url(default_storage.save(zip_filename, sio))})
 
+    def get_submission_download_status(self, student_id):
+        student_item_dict = self.student_submission_id(student_id)
+        staff_user = User.objects.get(id=self.xmodule_runtime.user_id)
```
Comment (marjev) [inline edx_sga/sga.py]: You can dry this into a property.
Then you can avoid doing this here and just send something like `self.user` when you call `get_download_status` or `set_as_downloaded`.

File: edx_sga/sga.py [hunk 9]
```diff
@@ -491,6 +495,16 @@ def download_zip(self, student_files):
         # save zip file and return its URL as JSON response
         return Response(json={'zip_url': default_storage.url(default_storage.save(zip_filename, sio))})
 
+    def get_submission_download_status(self, student_id):
+        student_item_dict = self.student_submission_id(student_id)
```
Comment (marjev) [inline edx_sga/sga.py]: You don't need this line since you're not doing anything with this variable except passing it to the other method.
You can simply call `student_submission_id` as an argument.

File: edx_sga/sga.py [hunk 10]
```diff
@@ -491,6 +495,16 @@ def download_zip(self, student_files):
...
+        staff_user = User.objects.get(id=self.xmodule_runtime.user_id)
+        return submissions_api.get_download_status(student_item_dict, staff_user)
+
+    def set_submission_as_downloaded(self, student_id):
+        student_item_dict = self.student_submission_id(student_id)
+        staff_user = User.objects.get(id=self.xmodule_runtime.user_id)
```
Comment (marjev) [inline edx_sga/sga.py]: See the comments for the method above.

File: edx_sga/static/js/src/edx_sga.js (hunk 3 above)
Comment (marjev) [reply edx_sga/static/js/src/edx_sga.js]: @iivic dafuq? the first comment

File: edx_sga/static/js/src/edx_sga.js (hunk 2 above)
Comment (marjev) [reply edx_sga/static/js/src/edx_sga.js]: I think you don't need this. You can always have this in the DOM and just toggle hide / show, right?

File: edx_sga/static/js/src/edx_sga.js (hunk 4 above)
Comment (marjev) [reply edx_sga/static/js/src/edx_sga.js]: To avoid doing a shamanistic ritual code, you can just add an extra class to the DOM element, like `.downloaded-indicator` or something like that and just toggle it (hide/show) when needed.

File: edx_sga/static/js/src/edx_sga.js [hunk 11]
```diff
@@ -161,13 +170,22 @@ function StaffGradedAssignmentXBlock(runtime, element) {
             $("button#download-selected-submissions").on("click", function() {
                 var student_ids = [];
                 $submissionCheckboxes.filter(":checked").each(function() {
+                    var $submissionFilename = $(this).closest("tr").find("a.submission-filename");
+                    if (!$submissionFilename.prev("i").length) {
+                        $submissionFilename.before(submissionDownloadedHtml);
+                    }
```
Comment (marjev) [inline edx_sga/static/js/src/edx_sga.js]: This looks very similar to the thing you're doing in the code above. You can create a function to DRY.

File: edx_sga/static/js/src/edx_sga.js [hunk 12]
```diff
@@ -161,13 +170,22 @@ function StaffGradedAssignmentXBlock(runtime, element) {
...
                 $.post(downloadSubmissionsUrl, JSON.stringify({'student_ids': student_ids})).done(function(data) {
                     window.location = window.location.origin + data.zip_url;
                 });
             });
             $("button#download-all-submissions").on("click", function() {
+                $("a.submission-filename").each(function() {
+                    if (!$(this).prev("i").length) {
+                       $(this).before(submissionDownloadedHtml);
+                    }
+                });
```
Comment (marjev) [inline edx_sga/static/js/src/edx_sga.js]: Same comment as above.

File: edx_sga/static/js/src/edx_sga.js (hunk 2 above)
Comment (marjev) [reply edx_sga/static/js/src/edx_sga.js]: Generally speaking, it's a bad idea to have a template in JS code (if you don't really have to).
It's to satisfy the principle called `Separation of concerns`.
You want the template to be in one file where you can clearly see the structure, and achieve the dynamics using the JS code (scripts). But, changing the tree structure is a bad idea most of the time because you might end up with the bug that's hard to handle.
If the template is as static as possible, the only thing you have to worry about is hiding and showing certain parts of the tree.

**Review by marjev — CHANGES_REQUESTED** [review]
Some cleanup can be done, but it seems to work.

Let me know once the comments are addressed so that I can take another look.

File: edx_sga/static/js/src/edx_sga.js (hunk 3 above)
Comment (iivic) [reply edx_sga/static/js/src/edx_sga.js]: It's good practice to keep selectors simple. Thus, the selector shouldn't contain unnecessary chained elements. If you think the tag is needed you should keep it of course.

File: edx_sga/sga.py (hunk 1 above)
Comment (ghost) [reply edx_sga/sga.py]: To make it user friendly. Currently the datetime is displayed like this: 2017-12-19T14:39:19.302707

File: edx_sga/static/js/src/edx_sga.js (hunk 2 above)
Comment (ghost) [reply edx_sga/static/js/src/edx_sga.js]: Yep, you have a good point, the show/hide approach makes sense.

**Review by marjev — APPROVED** [review]

**Review by iivic — APPROVED** [review]
